### ISSUE TITLE
OS X: Drive root error when run from Finder

### ISSUE THREAD

**Issue by chief1983** [scp-fs2open/fs2open.github.com#967]
![screenshot 2016-09-12 10 18 04](https://cloud.githubusercontent.com/assets/145825/18441316/3dec5e76-78d2-11e6-8e6e-741a0cd6e621.png)

I get this error when double-clicking a recent nightly installed into my FS2 directory.

**Comment by chief1983** [comment]
For those reading emails, the error is:  I get this error when double-clicking a recent nightly installed into my FS2 directory.

**Comment by SamuelCho** [comment]
So initial testing tells me that this has something to do with the g-buffers. Deferred lighting doesn't work on Mac but disabling it should look okay. I'd say just disable deferred lighting as a temporary workaround for now until I can find a solution. Thanks again for unblocking me chief.

**Comment by asarium** [comment]
Is this still an issue?

**Comment by chief1983** [comment]
Not sure, let me try the most recent nightly and report back.

**Comment by chief1983** [comment]
Yup, still get that error.

**Comment by asarium** [comment]
Can you step through `cfile_in_root_dir` in a debugger and determine why `cfile_in_root_dir` fails? Or you could add debug prints to determine what `buf` contains.

**Comment by chief1983** [comment]
If I could figure out how to get debugging to work on a Mac I could but I still haven't had much luck in that department yet, or we'd be a lot closer to fixing that sound crash too :(

**Comment by asarium** [comment]
Unfortunately, I cannot help you with that since I don't have a Mac. However, I can help you if you have questions about the CMake setup.

**Comment by jg18** [comment]
I think I tracked this down.

Apparently this was a known issue and there was an attempt to fix it in actual_main() freespace.cpp. There were a couple issues:

(1) The relevant code block checked for the APPLE_APP preprocessor macro, which AFAICT is not defined at least on the Xcode I'm using (7.2.1). Changing APPLE_APP to __APPLE__ fixes this.

(2) The code uses the function SetCurrentDirectory() which is not in the macOS APIs AFAICT. Switching to the POSIX equivalent chdir() fixes this.

With these changes, FSO now runs when run from the Finder.


The fix brings up a couple of other issues:

- The APPLE_APP macro is also used in os_init_cmdline() in cmdline.cpp. Through a similar fix, I've replaced the GetCurrentDIrectory() function with the POSIX function getcwd().

- The APPLE_APP macros is also used for #including #include <CoreFoundation/CoreFoundation.h> in cmdline.cpp and stubs.cpp. Given the APPLE_APP issue described above, this file is never #included, so maybe we don't need it?

Once question (2) is resolved, I can generate  a PR with the fixes.

**Comment by chief1983** [comment]
I have a feeling it was used because theoretically, FSO could be compiled and ran as a non-app binary, without being bundled or whatever.  To fix it we probably just need to have CMake define APPLE_APP whenever the Xcode generator is used?  If it uses a different generator I don't think bundling would be occurring.  Or if cmake is already telling the project file to make an app bundle somewhere, setting that project define should be done there.

**Comment by asarium** [comment]
We already have a platform detection system in place in our CMake setup which could be used to supply that definition. You could add a `target_compile_definitions` directive in [platform-darwin.cmake](https://github.com/scp-fs2open/fs2open.github.com/blob/master/cmake/platform-darwin.cmake) which defines the correct preprocessor symbol and then that definition should be available in all files of the project.

**Comment by chief1983** [comment]
Right, that could probably work if we want it enabled for all Apple builds, but what I was getting at is all Apple builds aren't necessarily app bundles are they?  If that's all we support, well ok, but it doesn't necessarily have to be all we support with cmake's flexibility.

**Comment by asarium** [comment]
At the moment, we only support bundle builds. If non-bundle builds ever become a thing we could add another preprocessor directive that enabled code that is specific to the bundles but I don't think we need to worry about that at the moment since the CMake system is currently designed to produce bundles.

**Comment by chief1983** [comment]
I guess it didn't seem that way, because if I didn't specify xcode on the cmake generator option, I thought I ended up with a gcc or clang build that wasn't bundled.  But I haven't tried a non-xcode generator in some time.

**Comment by asarium** [comment]
I don't think that will work at the moment since CMake currently assumes that MacOS implies a bundle build.

**Comment by chief1983** [comment]
Cool, if that's the case then throwing it in the platform specific code would work fine.  Just something to remember later if we ever want to undo that.

**Comment by chief1983** [comment]
One concern we might have is, will enabling APPLE_APP at this point have other unintended consequences?  Some of these ifdefs that looked for it might not be necessary anymore.  One even lives within our libjpeg configuration header.  If jpeg is working fine without it, would we still need that special behavior?  Might be a chance to reduce some code complexity here.

**Comment by chief1983** [comment]
I want to add more to this but I'm not sure where to begin.  As for CoreFoundation, I have no idea if it's used by something or not.  Probably not relelated to the current directory stuff since the SetCurrentDirectory call doesn't use the file.

I reduced the cmdline.cpp stuff down to one line:
```fp = fopen("data" DIR_SEPARATOR_STR "cmdline_fso.cfg", "rt");```
It may not be enough, and a getcwd/chdir combo may still be necessary for OS X when APPLE_APP is defined.  But then at least the WIN32 part can be removed and combined with the rest of the *nix else part.

The APPLE_APP in the jpeg config may not be necessary.  It hasn't been active for months.  Although, I haven't probably used a jpeg texture in a very long time.  Jpegs on Macs should probably be investigated to see if there are any issues happening with them.

The one in stubs may not be required any more either.  I'm wondering if Apple moved some things from CoreFoundation into more native libraries, or if we refactored out the need for it and didn't remove the include.

**Comment by chief1983** [comment]
This is what I have so far that seems to fix the root directory issue.  I still see some issues, one of which I just filed a bug for (no intro videos), but a couple which are not really deal breaks but still odd behavior.  One, when I copy the app into the game folder, the icon in Finder usually changes to a generic icon.  If I rename it on the command line, Finder won't even let me run it, claiming it is damaged.  I can rename 3.7.4 builds all day long, copy them, etc, and no oddness happens with the icons or trying to run them.  But this fixes the double click issue when it hasn't been renamed, so it is probably enough for this ticket.  Still not sure about whether CoreFoundation is needed, I tried with and without that include and I couldn't see a behavior change (never did actually enter the game though).

```
diff --git a/cmake/platform-darwin.cmake b/cmake/platform-darwin.cmake
index 2b21f1d..3b1db60 100644
--- a/cmake/platform-darwin.cmake
+++ b/cmake/platform-darwin.cmake
@@ -1,5 +1,6 @@
 
 MESSAGE(STATUS "Doing Mac OSX specific things...")
+target_compile_definitions(platform INTERFACE APPLE_APP)
 SET(EXE_GUI_TYPE MACOSX_BUNDLE)
 
 FIND_LIBRARY(COCOA_LIBRARY Cocoa)
diff --git a/code/cmdline/cmdline.cpp b/code/cmdline/cmdline.cpp
index 58922e3..e439c70 100644
--- a/code/cmdline/cmdline.cpp
+++ b/code/cmdline/cmdline.cpp
@@ -954,17 +954,15 @@ void os_init_cmdline(int argc, char *argv[])
                        // read the cmdline_fso.cfg file from the data folder, and pass the command line arguments to
                        // the the parse_parms and validate_parms line.  Read these first so anything actually on
                        // the command line will take precedence
-#ifdef _WIN32
-                       fp = fopen("data\\cmdline_fso.cfg", "rt");
-#elif defined(APPLE_APP)
+#if defined(APPLE_APP)
                        char resolved_path[MAX_PATH], data_path[MAX_PATH_LEN];
 
-                       GetCurrentDirectory(MAX_PATH_LEN - 1, data_path);
+                       getcwd(data_path, sizeof(data_path));
                        snprintf(resolved_path, MAX_PATH, "%s/data/cmdline_fso.cfg", data_path);
 
                        fp = fopen(resolved_path, "rt");
 #else
-                       fp = fopen("data/cmdline_fso.cfg", "rt");
+                       fp = fopen("data" DIR_SEPARATOR_STR "cmdline_fso.cfg", "rt");
 #endif
 
                        // if the file exists, get a single line, and deal with it
diff --git a/freespace2/freespace.cpp b/freespace2/freespace.cpp
index 43f8449..e69a8c3 100644
--- a/freespace2/freespace.cpp
+++ b/freespace2/freespace.cpp
@@ -8092,7 +8092,7 @@ int actual_main(int argc, char *argv[])
        // Finder sets the working directory to the root of the drive so we have to get a little creative
        // to find out where on the disk we should be running from for CFILE's sake.
        char *path_name = SDL_GetBasePath();
-       SetCurrentDirectory(path_name);
+       chdir(path_name);
        SDL_free(path_name);
 #endif
 
```

**Comment by asarium** [comment]
The diff looks good (maybe you could add a few blank lines for better readability in the CMake part :p) so you could probably submit that as a PR. That should fix this issue and IMO the other issues you found should be tracked in their own issues just to keep everything organized.

**Comment by chief1983** [comment]
i had kinda wanted that line to be grouped with the bundled line in case that ever gets optional in the future, for some reason.  And I will, just want to see if that agrees with what jg18 had in mind.

**Comment by jg18** [comment]
Maybe outside the scope of this issue, but shouldn't data_path and resolved_path from the modified part of cmdline.cpp be cleared out with a memset before use? And also shouldn't they be set to terminate with a null char (set last char to null char) just to be safe?

**Comment by chief1983** [comment]
I double checked, but most of the usage examples I can find online don't bother with that, when using the buffer argument.  I suppose it could be done but I think it's outside of the scope and my likelihood to get it right the first try to add it to this bugfix.  If we want discussion on that further, it can be in the PR I will go ahead and open.

**Comment by Goober5000** [comment]
Closing this since #1227 was merged.
